### PR TITLE
feat(pii): implement a masking strategy for UPI VPAs

### LINKED ISSUE #1578: [FEATURE] Implement a masking strategy for UPI VPAs
### Feature Description

Unified Payments Interface (UPI) is a popular payment technique in India, used to transfer money among individuals, bill payments and merchant payments. In the case of money transfer among individuals, users are identified with a Virtual Payment Address (VPA). A VPA is typically of the following format: `<identifier>@<bank_name>`. (Feel free to read up more about UPI if you aren't already familiar with it.)

Since a VPA can uniquely identify a user, we'd want to mask VPAs in our application. However, we'd want to keep the `<bank_name>` part unmasked, since it could help us build analytics on the banks or payment service providers used, and possibly improve our routing experience.

### Possible Implementation

We'd want you to implement `masking::Strategy` for UPI VPAs (which could be called `UpiVpaMaskingStrategy`) which masks the identifier completely, and keeps the bank name part in clear text. You can refer to the email masking strategy to get an idea of how this can be done. Please include unit tests for the same.

https://github.com/juspay/hyperswitch/blob/2f9c28938f95a58532604817b1ed370ef8285dd8/crates/common_utils/src/pii.rs#L190-L205

Also, update the type of `vpa_id` field in `api_models::payments::UpiData` to be `Option<Secret<String, UpiVpaMaskingStrategy>>` instead.

https://github.com/juspay/hyperswitch/blob/2f9c28938f95a58532604817b1ed370ef8285dd8/crates/api_models/src/payments.rs#L783-L786

### PR DESCRIPTION
**Pull request by SargamPuram**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Here, I have implemented masking::Strategy for UPI VPAs which is called UpiVpaMaskingStrategy which masks the identifier completely, and keeps the bank name part in clear text. I have included tests for the same. 

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Fixes #1578.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by SargamPuram** [comment]
@SanchithHegde , I have committed the suggested changes . Let me know if any more changes are needed. Thanks for guiding me!

File: crates/common_utils/src/pii.rs [hunk 1]
```diff
@@ -329,13 +329,33 @@ where
...
+where
+    T: AsRef<str> + std::fmt::Debug,
+{
+    fn fmt(val: &T, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let vpa_str: &str = val.as_ref();
+        if let Some((user_identifier, bank_or_psp)) = vpa_str.split_once('@') {
+            let masked_user_identifier = "*".repeat(user_identifier.len());
+            write!(f, "{masked_user_identifier}@{bank_or_psp}");
+        } else {
+            WithType::fmt(val, f),
```
Comment (SanchithHegde) [inline crates/common_utils/src/pii.rs:347]:
```suggestion
            write!(f, "{masked_user_identifier}@{bank_or_psp}")
        } else {
            WithType::fmt(val, f)
```

Minor changes to fix formatting and failing CI checks.

Comment (SargamPuram) [reply crates/common_utils/src/pii.rs:347]: Yes, I'll commit it .

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -834,7 +834,7 @@ pub struct CryptoData {
 #[serde(rename_all = "snake_case")]
 pub struct UpiData {
     #[schema(value_type = Option<String>, example = "successtest@iata")]
-    pub vpa_id: Option<Secret<String>>,
+    pub vpa_id: Option<Secret<String, UpiVpaMaskingStrategy>>,
```
[the remark this answers is not in the document]
Comment (SanchithHegde) [reply crates/api_models/src/payments.rs]: My bad, I missed out an extra `:`.

**Comment by ashokkjag** [comment]
Just curious on why not use email strategy?

**Comment by SanchithHegde** [comment]
@SargamPuram To fix the failing CI checks, could you please add this snippet in the `crates/masking/src/abs.rs` file:

```rust
/// Interface that consumes a secret and converts it to a secret with a different masking strategy.
pub trait SwitchStrategy<FromStrategy, ToStrategy> {
    /// The type returned by `switch_strategy()`.
    type Output;

    /// Consumes the secret and converts it to a secret with a different masking strategy.
    fn switch_strategy(self) -> Self::Output;
}

impl<S, FromStrategy, ToStrategy> SwitchStrategy<FromStrategy, ToStrategy>
    for Secret<S, FromStrategy>
where
    FromStrategy: crate::Strategy<S>,
    ToStrategy: crate::Strategy<S>,
{
    type Output = Secret<S, ToStrategy>;

    fn switch_strategy(self) -> Self::Output {
        Secret::new(self.inner_secret)
    }
}
```

Once that's done, could you update this line:

https://github.com/juspay/hyperswitch/blob/f77fdb7a6ed354151d8a758a734382a4c3b2698e/crates/router/src/connector/iatapay/transformers.rs#L88

to be (you'll also have to import the `SwitchStrategy` trait):

```upi_data.vpa_id.map(|id| PayerInfo { token_id: id.switch_strategy() })```

**Comment by SanchithHegde** [comment]
> Just curious on why not use email strategy?

We could end up having different strategies for email and VPAs on a later date if needed: say we decide to unmask the first 2 characters of either email or VPA.

**Comment by SargamPuram** [comment]
> @SargamPuram To fix the failing CI checks, could you please add this snippet in the `crates/masking/src/abs.rs` file:
> 
> ```rust
> /// Interface that consumes a secret and converts it to a secret with a different masking strategy.
> pub trait SwitchStrategy<FromStrategy, ToStrategy> {
>     /// The type returned by `switch_strategy()`.
>     type Output;
> 
>     /// Consumes the secret and converts it to a secret with a different masking strategy.
>     fn switch_strategy(self) -> Self::Output;
> }
> 
> impl<S, FromStrategy, ToStrategy> SwitchStrategy<FromStrategy, ToStrategy>
>     for Secret<S, FromStrategy>
> where
>     FromStrategy: crate::Strategy<S>,
>     ToStrategy: crate::Strategy<S>,
> {
>     type Output = Secret<S, ToStrategy>;
> 
>     fn switch_strategy(self) -> Self::Output {
>         Secret::new(self.inner_secret)
>     }
> }
> ```
> 
> Once that's done, could you update this line:
> 
> https://github.com/juspay/hyperswitch/blob/f77fdb7a6ed354151d8a758a734382a4c3b2698e/crates/router/src/connector/iatapay/transformers.rs#L88
> 
> to be (you'll also have to import the `SwitchStrategy` trait):
> 
> `upi_data.vpa_id.map(|id| PayerInfo { token_id: id.switch_strategy() })`

@SanchithHegde , yes I'll do that .

**Review by SanchithHegde — DISMISSED** [review]

**Comment by SanchithHegde** [comment]
> should i add use masking::abs::SwitchStrategy; or use masking::{Secret, abs::SwitchStrategy}; to import the trait so that it will fix CI checks?

Yes, please address the failing CI checks.

**Comment by SanchithHegde** [comment]
@SargamPuram Let me know if you need any help with fixing the failing CI checks.

**Comment by SargamPuram** [comment]
> @SargamPuram Let me know if you need any help with fixing the failing CI checks.

@SanchithHegde , yes , please ! I could really use some help with fixing the failing CI checks.
Sorry for the delay from my end , I was occupied with university examinations .

**Comment by prasunna09** [comment]
> @SanchithHegde , yes , please ! I could really use some help with fixing the failing CI checks. Sorry for the delay from my end , I was occupied with university examinations .
 
@SargamPuram Can you please run the following commands - 
 - `cargo +nightly fmt` (to fix the formatting check)
 - "`use masking::abs::SwitchStrategy;`" Please import this "SwitchStrategy" to fix clippy error
 This should fix all the checks.
 Also, you can use this command in local - "`cargo clippy --all-features --all-targets`"

**Comment by SargamPuram** [comment]
> > @SanchithHegde , yes , please ! I could really use some help with fixing the failing CI checks. Sorry for the delay from my end , I was occupied with university examinations .
> 
> @SargamPuram Can you please run the following commands -
> 
> * `cargo +nightly fmt` (to fix the formatting check)
> * "`use masking::abs::SwitchStrategy;`" Please import this "SwitchStrategy" to fix clippy error
>   This should fix all the checks.
>   Also, you can use this command in local - "`cargo clippy --all-features --all-targets`"

@prasunna09 , yes I'll run the cargo +nightly fmt for the formatting checks.
As for the use masking::abs::SwitchStrategy; , which you have suggested , I did try to import SwitchStrategy using this , in one of the previous commits . But it displayed this error
juspay5
I suspect that by again using  use masking::abs::SwitchStrategy; , it'll display the same error , can you suggest some other way to import the SwitchStrategy trait , so that'll help fix the CI checks.

**Comment by SanchithHegde** [comment]
@SargamPuram Can you re-export `SwitchStrategy` on this line:

https://github.com/SargamPuram/hyperswitch/blob/823dbc92beb044ebfd080bfaafe6c455c7403ab8/crates/masking/src/lib.rs#L19

Like so:

```rust
pub use abs::{ExposeInterface, ExposeOptionInterface, PeekInterface, SwitchStrategy};
```

**Comment by SargamPuram** [comment]
> masking::SwitchStrategy

@SanchithHegde , hey !
I've made all the changes that were suggested . 
1.Re-exported the SwitchStrategy in the specified line . 
exportedswitchstrategy
2.Imported masking::SwitchStrategy instead of masking::abs::SwitchStrategy.
importedSwitchStrategy
3. Run the cargo +nightly fmt for fixing formatting checks.
cargonightlyfmt
But , I'm not able to push these commits from local , I'm getting this error message 
errormessage
I've tried git pull and my local repository is up-to-date with the remote. 
Will it be okay if i close this PR and create a new PR with all the changes from the start , hopefully that will pass all the CI checks .

**Comment by SanchithHegde** [comment]
> I'm not able to push these commits from local , I'm getting this error message

Can you try running `git pull --rebase` and then run `git push`?

**Comment by SargamPuram** [comment]
> > I'm not able to push these commits from local , I'm getting this error message
> 
> Can you try running `git pull --rebase` and then run `git push`?

Yes, I've tried that too . Still it keeps showing me the same error . 
will it be okay if i would create a new PR and close this one?

**Comment by SanchithHegde** [comment]
> Yes, I've tried that too . Still it keeps showing me the same error . will it be okay if i would create a new PR and close this one?

We'd prefer using this PR itself since it would include all previous conversation. Could you try pushing it to a different branch or force pushing to this branch?

**Comment by SargamPuram** [comment]
@prasunna09 thanks a lot , for helping in fixing the CI checks .

**Review by SanchithHegde — DISMISSED** [review]

**Review by SanchithHegde — DISMISSED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**